external-crates/move/crates/move-regex-borrow-graph/src/collections.rs

```diff
@@ -0,0 +1,510 @@
+        for (r, other_node) in other.nodes.iter().filter(|(r, _)| self_keys.contains(r)) {
+            let self_node = self.node_mut(r)?;
+            size_increase = size_increase.saturating_add(self_node.join(&self_keys, other_node));
+        }
+        self.abstract_size = self.abstract_size.saturating_add(size_increase);
+        self.check_invariant();
+        Ok(size_increase > 0)
+    }
+
+    pub fn refresh_refs(&mut self) -> Result<()> {
+        let nodes = std::mem::take(&mut self.nodes);
+        self.fresh_id = 0;
+        self.nodes = nodes
+            .into_iter()
+            .map(|(r, node)| {
+                let r = r.refresh()?;
+                self.fresh_id = std::cmp::max(self.fresh_id, r.fresh_id()? + 1);
+                let node = node.refresh_refs()?;
+                Ok((r, node))
+            })
+            .collect::<Result<_>>()?;
+        debug_assert!(self.is_fresh());
+        Ok(())
+    }
+
+    pub fn canonicalize(&mut self, remapping: &BTreeMap<Ref, usize>) -> Result<()> {
+        let nodes = std::mem::take(&mut self.nodes);
+        self.nodes = nodes
+            .into_iter()
+            .map(|(r, node)| Ok((r.canonicalize(remapping)?, node.canonicalize(remapping)?)))
+            .collect::<Result<_>>()?;
+        self.fresh_id = 0;
+        debug_assert!(self.is_canonical());
+        Ok(())
+    }
+
+    pub fn is_canonical(&self) -> bool {
+        self.nodes.keys().all(|r| r.is_canonical())
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        self.nodes.keys().all(|r| r.is_fresh())
+    }
+
+    //**********************************************************************************************
+    // Invariants
+    //**********************************************************************************************
+
+    // checks:
+    // - both graphs satisfy their invariants
+    // - all nodes are canonical
+    // - all nodes in self are also in other
+    fn check_join_invariants(&self, other: &Self) {
+        #[cfg(debug_assertions)]
+        {
+            self.check_invariant();
+            other.check_invariant();
+            for self_r in self.keys() {
+                debug_assert!(self_r.is_canonical());
+                debug_assert!(other.nodes.contains_key(&self_r));
+            }
+            for other_r in other.keys() {
+                debug_assert!(other_r.is_canonical());
+                // there can be nodes in other that are not in self
+            }
+            for (self_r, self_node) in &self.nodes {
+                let other_node = other.node(self_r).unwrap();
+                debug_assert_eq!(self_node.ref_(), other_node.ref_());
+                debug_assert_eq!(self_node.is_mutable(), other_node.is_mutable());
+            }
+        }
+    }
+
+    // checks:
+    // - ref --> node has ref == node.ref()
+    // - successor/predecessor relationship is correctly maintained
+    // - the abstract size is correct
+    pub fn check_invariant(&self) {
```

Comment:
Can we rename this? Inline, it looks like it is enforcing something about borrowing, not graph structural assertions.